train.py

```diff
@@ -0,0 +1,207 @@
+        loss = self.loss_function_module(predicted_noise, added_noise)
+
+        self.log("loss", loss, on_epoch=True)
+
+        if self.global_step % self.hparams["save_every_x_steps"] == 0:
+            self._save_models()
+
+        return loss
+    
+    def _save_models(self):
+        # ckpt specifies directory and name of the file is name of the experiment in wandb
+        save_path = f"{self.hparams['ckpt_dir']}/{self.logger.experiment.name}.ckpt"
+        # saving models
+        torch.save(
+            {
+                "model": self.model.state_dict(),
+                "forward_diffusion": self.forward_diffusion.state_dict()
+            },
+            save_path
+        )
+
+
+def preprocess_dataset(filepath: str, batch_size: int):
+    dataset = load_dataset(filepath)
+    transformed_dataset = dataset.with_transform(transforms).remove_columns("label")
+
+    dataloader = DataLoader(transformed_dataset["train"], batch_size=batch_size)
+
+    return dataloader
+
+def transforms(examples):
+    transform = Transforms(img_size=28)
+
+    examples["x"] = [transform.img2torch(image.convert("L")) for image in examples["image"]]
+    del examples["image"]
+
+    return examples
+
+
+def train(dataloader: DataLoader, args):
+    # generate some random id for the run
+    run_id = str(uuid.uuid1())[:8]
+
+    # initializing logger
+    logger = WandbLogger(project="annotated-diffusion", name=f"{args.run_name}-{run_id}", save_dir=args.log_dir)
+
+    # initializing models
+    model = Unet(
+        dim=args.dim, dim_mults=args.dim_mults,
+        in_channels=args.in_channels, resnet_block_groups=args.num_resnet_groups
+    )
+
+    forward_diffusion = ForwardDiffusion(
+        beta_start=args.beta_start, beta_end=args.beta_end, 
+        timesteps=args.timesteps, schedule_type=args.schedule_type
+    )
+
+    # if checkpoint exists load state dict for model and forward diffusion
+    if args.load_ckpt_path is not None:
+        checkpoint = torch.load(args.load_ckpt_path)
+        model.load_state_dict(checkpoint["model"])
+        forward_diffusion.load_state_dict(checkpoint["forward_diffusion"])
+
+    # initializng lightning model wrapper that will handle training logic
+    model_training_wrapper = UnetTrainingWrapper(
+        model, forward_diffusion, lr=args.lr, l2=args.l2, 
+        loss_type=args.loss_type, ckpt_dir=args.save_ckpt_dir, save_every_x_steps=args.save_every_x_steps
+    )
+
+    # callbacks
+    callbacks = [TQDMProgressBar()]
+
+    # initializing trainer with specified hyperparameters
+    trainer = pl.Trainer(
+        logger=logger, callbacks=callbacks, max_epochs=args.num_epochs, 
+        accelerator=args.accelerator, precision=args.precision, 
+        log_every_n_steps=args.log_every_n_steps, overfit_batches=args.overfit_batches
+    )
+
+    # run training
+    trainer.fit(model_training_wrapper, dataloader)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(prog="annotated-diffusion")
```

Comment:
I'm not a huge fan of large numbers of parameters handled with `argparse`. For ML projects I've been recently experimenting with [hydra](https://hydra.cc/docs/intro/) which seems to work great to handle training related configs and stuff like that, please take a look!

Reply:
Ok, I'll check that